msg_sequence/test_pairing.py

```diff
@@ -4,98 +4,76 @@
+from helpers import chains, deproxy, dmesg
 
 __author__ = "Tempesta Technologies, Inc."
 __copyright__ = "Copyright (C) 2018 Tempesta Technologies, Inc."
 __license__ = "GPL2"
 
 
-class PairingTest(functional.FunctionalTest):
-    chains_size = 2
-    send_to_close = chains_size / 2
-    defconfig = ""
-
-    def create_servers(self):
-        port = tempesta.upstream_port_start_from()
-        self.servers = [deproxy.Server(port=port, conns_n=1)]
-
-    def create_tester(self):
-        message_chains = [chains.proxy() for _ in range(self.chains_size)]
-        self.tester = PairingTester(self.client, self.servers, message_chains=message_chains)
-
-    def prepare(self):
-        self.tempesta.config.set_defconfig(self.defconfig)
-        self.configure_tempesta()
-
-        for server in self.servers:
-            server.start()
-
-        self.tempesta.start()
-        self.client.start()
-        self.tester.start()
-
+class PairingTest(tester.TempestaTest):
+    backends = [
+        {
+            "id": "backend",
+            "type": "deproxy",
+            "port": "8000",
+            "response": "static",
+            "response_content": "HTTP/1.1 200 OK\r\n",
+        }
+    ]
+    clients = [
+        {
+            "id": "client", 
+            "type": "deproxy", 
+            "addr": "${tempesta_ip}", 
+            "port": "80",
+        },
+    ]
+    tempesta = {
+        "config": """
+        listen 80;
+        server ${server_ip}:8000;
+        cache 0;
+        """
+    }
+    
     @dmesg.unlimited_rate_on_tempesta_node
     def test_disconnect_client(self):
         """Tempesta forwards requests from client to backend, but client
         disconnects before Tempesta received responses from backend. Responses
         must be evicted, no 'Paired request missing' messages are allowed.
         """
-        self.prepare()
-        self.tester.loop(0.5)  # Let handle connects
-        self.tester.send_reqs(self.send_to_close)
-        self.tester.disconnect_clnt()
-        self.tester.send_resps()
+        chains_size = 2
+        send_to_close = chains_size / 2
+        message_chains = [chains.proxy() for _ in range(chains_size)]
+        message_chains_recv = [deproxy.MessageChain.empty() for _ in message_chains]
+        last_req = 0
+
+        self.start_all_services()
+        client = self.get_client("client")
+        server = self.get_server("backend")
+
+        def send_reqs(req_n: float) -> None:
+            e_req = last_req + req_n
+            b_req = last_req
+
+            for i in range(int(b_req), int(e_req)):
+                client.clear()
+                client.set_request(message_chains[i])
+
+        def send_resps() -> None:
+            conn = server.connections[0]
+            for i in range(last_req):
+                if message_chains_recv[i].fwd_request:
+                    conn.send_response(message_chains[i].server_response)
+                else:
+                    break
```

Comment:
You should not a connection from DeproxyServer. Tempesta creates 32 connections by default (see [conns_n](https://github.com/tempesta-tech/tempesta/wiki/Backend-servers) from `server` directive. You don't know which connection was received a request. Please make `connection` a private variable when you remove the `tester` framework.
And also in this test server's response is incomplete and Tempesta will wait for a second part of response. I think it should be used here:
```python3
from helpers import util

# wait until a server receives all requests
util.wait_until(lambda: len(server.requests) != chains_size)
# server received all requests, but Tempesta did not receive response so we just close connections
client.stop()
```